libs/core/langchain_core/vectorstores.py

```diff
@@ -213,6 +478,38 @@ async def aadd_documents(
         Returns:
             List of IDs of the added texts.
         """
+        # If either upsert or aupsert has been implemented, we delegate to them!
+        if (
+            type(self).aupsert != VectorStore.aupsert
+            or type(self).upsert != VectorStore.upsert
+        ):
+            # If aupsert has been implemented, we can use it to add documents
+            from langchain_core.documents import Document
+
+            if "ids" in kwargs:
+                ids = kwargs.pop("ids")
+                if ids and len(ids) != len(documents):
+                    raise ValueError(
+                        "The number of ids must match the number of documents."
+                        "Got {len(ids)} ids and {len(documents)} documents."
+                    )
+
+                documents_ = []
+
+                for id_, document in zip(ids, documents):
+                    doc_with_id = Document(
+                        page_content=document.page_content,
+                        metadata=document.metadata,
+                        id=id_,
+                    )
+                    documents_.append(doc_with_id)
+            else:
+                documents_ = documents
+
+            # If upsert has been implemented, we can use it to add documents
```

Comment:
(nit) this comment is a little confusing as we only call `self.aupsert` (understood that `.aupsert` by default will use `.upsert`).

Reply:
Good point, I'll follow up with an update of the doc-string